README.md

Lenc
===========

LEnc is a Ruby gem that maintains encrypted repositories of files, enabling secure, encrypted
backups to free cloud services such as Dropbox, Google Drive, and Microsoft SkyDrive.

Written by Jeff Sember, March 2013.

[Source code documentation can be found here.](http://rubydoc.info/gems/lenc/frames)


Directory structure
--------
The program manipulates three distinct directories:

* A \<source\> directory, which holds all the
files to be encrypted.  The only file that the program modifies within this
directory tree is a hidden configuration file ".lenc" (on Windows, this file
is named "__lenc_repo__.txt").

* An \<encrypted\> directory, where the program stores the encrypted
versions of all the files found in the <source> directory.  This directory
is usually mapped to a cloud service (e.g., Dropbox), or perhaps to a thumb drive.
		
		NOTE:  THE <encrypted> DIRECTORY IS MANAGED BY THE PROGRAM!  
		ANY FILES WRITTEN TO THIS DIRECTORY BY THE USER MAY BE DELETED.

* A \<recover\> directory.  The program can recover a set of encrypted files here.
For safety, the this directory must not lie within an existing repository.


Running the program
----

The program can be asked to perform one of the following tasks:

__Setting up a repository.__  Select a directory you wish to be the \<source\> 
directory, and make it the current directory.  Type:
    
		lencrypt -i KEY ENCDIR
    
with KEY a set of characters (8 to 56 letters) to be used as the encryption key,
and ENCDIR the name of the \<encrypted\> directory (it must not already exist, and
it cannot lie within the current directory's tree).
 

__Updating a repository.__ From within a \<source\> directory tree, type:

		lencrypt
    
You will be prompted for the encryption key, and then the program will examine 
which files within the \<source\> directory have been changed (since the repository 
was created or last updated), and re-encrypt these into the \<encrypted\> directory.


__Recovering encrypted files.__  Type:

		lencrypt -r KEY ENCDIR RECDIR
    
with KEY the key associated with a repository whose encrypted files are stored in ENCDIR.
The recovered files will be stored in RECDIR.


Additional options can be found by typing:

		lencrypt -h
    
    
Ignore files
----------------
If desired, you can avoid storing selected files in the encryption repository.  
Within the \<source\> directory (or any of its subdirectories), place a text 
file '.lencignore' with a list of file or directory names (or patterns) to be 
ignored.  Example:

    # This is a comment
    #
    log
    *.mp3
    _SKIP_*
   
This causes the program to ignore any file or directory named 'log', as well as
any ending with ".mp3" or starting with "_SKIP_".

Some files are automatically ignored, e.g. ".DS_Store". 

The format of ignore files is similar to that of .gitignore files.  Details:
 
* Each line should contain a single pattern representing files or directories to be ignored.
* A line will be ignored (treated as a comment) if it is blank, or if it starts with '#'.
* The path separator should be '/' (Mac, Unix) or '\' (Windows).
* If a pattern starts with '#', you can precede it with '\' to avoid it being ignored.
* Precede a pattern with '!' to specifically include files/directories, overriding any previous
 		matching pattern in a parent directory.
* If a pattern ends with the path separator, it will be removed, and the pattern will
 		match only directories, not files.
* The wildcard '*' matches for any sequence of zero or more characters.
* The wildcard '?' matches any single character.
* If the pattern contains any path separators, then the wildcards '*', '?' will not
 		match the path separator.
 
Miscellaneous Issues
-------
* At present, the program will ignore any files (or directories) that are symbolic links.
 


CHANGELOG.txt
2013-03-21 
  * Version 1.0.0 released

2013-03-22 
  * 1.1.0 released
  * For added safety, unless the --storekey option is given when the repository is
    initialized, the key is no longer stored in repository configuration file, and
    the user is prompted at every update to enter the password.

2013-04-08
  * 1.1.1 released
	* Fixed problem with README that was omitting a word
	* Now allows passwords as short as 8 characters, to agree with documentation
	* Now omitting symlink'd files (it used to fail upon encountering these)
	* Added ability to provide update password as command line argument (i.e. lencrypt <pwd>)

2013-04-09
  * 1.1.2  released
  * Now hosting on GitHub
  * Updated README file

2013-04-15
  * No longer updating CHANGELOG; see git commit comments instead.  


test/test.rb
require 'test/unit'

require_relative '../lib/lenc/tools.rb'
req('repo lencrypt')


#SINGLETEST = "test_101_update_repo_with_verify"
if defined? SINGLETEST
  if main?(__FILE__)
    ARGV.concat("-n  #{SINGLETEST}".split)
  end
end

class RepoTest < MyTestSuite

  include LEnc
  
  # Make a directory, if it doesn't already exist
  def mkdir(name)
    if !File.directory?(name)  
      Dir::mkdir(name)
    end
  end

  def suite_setup
    #    pr("\n\n>>>> RepoTest   setup\n\n")
    
    # Make current directory = the one containing this script
    main?(__FILE__)
    
    @@testDir = "__temp_dirs__"
    mkdir(@@testDir)
      
    @@sourceDir = File.join(@@testDir,"__source__")
    @@encryptDir = File.join(@@testDir,"__encrypted__")
    @@recoverDir = File.join(@@testDir,"__recovered__")
    @@repoFile = File.join(@@sourceDir,Repo::LENC_REPO_FILENAME)
    @@key = "onefishtwofishredfishbluefish"
    @@sampleFile = "demeter"
    
    if !File.directory?(@@sourceDir)
      create_source_tree()
    end
    
    clean()
    
    # Construct a list of the source subdirectories, since some of them have
    # unicode filenames and OSX changes them on us.
    @@sourceDirs = []
    getSourceDirList(@@sourceDirs, @@sourceDir)
  end

    
  def getSourceDirList(lst, dr)  
    dir_entries(dr).each do |f|
      child = File.join(dr,f)
      if File.directory? child
        lst.push child
        getSourceDirList(lst, child)
      end
    end
  end
  
    
  # Delete .lenc file from source directory,
  # and delete the encryption and recovery directories
  def clean
#    pr("clean, repoFile=%s exists=%s\n",@@repoFile,File.file?(@@repoFile))
    remove_file_or_dir(@@repoFile)
    remove_file_or_dir(@@encryptDir)
    remove_file_or_dir(@@recoverDir)
#    remove_file_or_dir(bogusSourceFile())
  end
  
  
  def suite_teardown
#    pr("\n\n<<<< RepoTest   teardown\n\n")
    
    if false
      warn("always removing source dir")
      
      if @@sourceDir.end_with?("__source__")
        pr("removing...\n")
        remove_file_or_dir(@@sourceDir)
      end
    end
  end
  
  def method_setup
#    pr("\n\n\n")
  end
  
  def method_teardown
#    pr("\n\n\n")
  end

      
    
  def makeFile(pth, lngth=3)  
    remove_file_or_dir(pth)
    
    srand(1965)
    
    File.open(pth, 'wb')  do |f|
      j = 0
      while j < lngth  
        if (j+1) % 40 == 0
            f.write("\n")
        else
            f.write((rand(26)+65).chr)
        end
        j += 1
      end
    end
  end
     
  def unicode(c)
    s = [c].pack("U*")
    s
  end
    
  def create_source_tree  
    # Create a set of directories and files within the repo's source directory.
    
    # Specify a script as a sequence of names, each followed by
    #  | : file
    #  / : directory
    #  ^ : parent directory (name should be empty)
    
    # Some names have unicode characters for test purposes.
    
    scr =<<SCR
   
    mars| sol| luna| aphrodite | apollo | ares | artemis | hades | demeter |
    alpha/ 
       earth| venus| symphony 1 | symphony 2 | hephaestus | hera | hermes | hestia |
       delt#{unicode(0xe5)} / 
          jupiter| saturn| poseidon | zeus | 
       ^ 
    ^ 
    bet#{unicode(0xe1)} gamma|
      uranus|
      neptune|
      pluto|
SCR
      
      # Script of predetermined file lengths:
    cs = RepoInternal::CHUNK_SIZE_ENCR
    hs = RepoInternal::CHUNK_HEADER_SIZE
    flens = [0, cs - 1, cs, cs + 1, cs + hs - 1, cs + hs, cs + hs + 1,
             (2 * cs) - 1, 2 * cs, 2 * cs + 1, 2 * (cs + hs) - 1, 2 * (cs + hs), 2 * (cs + hs) + 1]
    fNum = 0
    
    srand(1983)
    
    dr = @@sourceDir 
    mkdir(dr)
    
    dirStack = []
    k = 0
    k0 = k
    while k < scr.size
      c = scr[k]
      k += 1
      next if !"|/^".index(c)
      
      nm = scr[k0...k-1].strip
      
      k0 = k
      
      if c == '|'  
        # create a file; use our predetermined lengths, if we still have some;
        # otherwise, choose a random file length
        if fNum < flens.size
          j = flens[fNum]
        else  
          j = (rand() * rand() * 130000).to_i + 3
        end
        fNum += 1
        makeFile(File.join(dr, nm), j)
      elsif c == '/'
          dirStack.push(dr)
          dr = File.join(dr,nm)
          mkdir(dr)
elsif c == '^'
          dr = dirStack.pop
end
end

      # Store some .lencignore files
    dr = @@sourceDir
    writeIgnore(dr, "s*")
    
    dr = File.join(dr, "alpha")
    writeIgnore(dr, "hera")
    
    dr = File.join(dr, "delt#{unicode(0xe5)}")
    writeIgnore(dr, "!s*")
  end


  def writeIgnore(dr, contents)  
    pth = File.join(dr,Repo::IGNOREFILENAME)
    File.open(pth,'w') do |f|
      f.write("# This is a .lencignore file\n#{contents}\n\n### end ###\n")
    end
  end
      
  def ex(args)  
    if args.is_a? String
      args = args.split
    end
    args.concat(["-w", @@sourceDir, "-q"])
    LEncApp.new().run(args)
  end
  
#  def bogusSourceFile  
#    return File.join(@@sourceDir, "_#bogusfile.txt")
#  end

  def build_repo_obj(silent = true)
    v = silent ? -1 : 0
    rp = Repo.new(:verbosity => v)
    rp
  end
    
  def create_repo
    clean
    rp = build_repo_obj
    rp.create(@@sourceDir,@@key,@@encryptDir, true)
    rp.close  
    rp
  end
  
  def update_repo(silent = true)
    rp = build_repo_obj(silent)
    rp.open(@@sourceDir)
    rp.perform_update
    rp.close  
  end
  
  def do_recover
    rp = build_repo_obj
    rp.perform_recovery(@@key, @@encryptDir,@@recoverDir)
    rp.close
  end
  
  # --------------- tests --------------------------
  
  
  
  def test_050_AES_encryption
    req 'aes'
          
    f = RepoInternal::MyAES.new(true, @@key, "42")
          
    originalText = ""
    while originalText.size < 100000
      64.times do |i|
        ch = ((i % 26) + 65).chr
        originalText << ch * 17
      end
    end
    
    enc = ""
    s = originalText.size
    n = 0
    while n < s 
      c = [60,s-n].min
      f.add(originalText[n...n+c])
      n += c
      enc << f.flush()
    end
          
    f.finish()
    enc << f.flush
     
    require 'digest/md5'
  
    digest = Digest::MD5.hexdigest(enc)
    assert(digest.to_s == '284740fbf3355951e9e76fb43fda985c')

    f = RepoInternal::MyAES.new(false, @@key)
    dec = ""
    n = 0
    s = enc.size
    while n < s 
      c = [60,s-n].min
      f.add(enc[n...n+c])
      n += c
      dec << f.flush
    end
    
    f.finish()
    dec << f.flush
    assert(originalText == dec)
  end
  
  
  # Create repo using Repo class
  def test_100_create_repo
    create_repo
    assert(File.file?(@@repoFile))
  end
  
  # Create repo using lencrypt program
  def test_101_create_repo
    clean
    a = "-i #{@@key} #{@@encryptDir} --storekey"
    ex(a)
  end

  def test_101_update_repo_with_verify
    clean
    ex("-i #{@@key} #{@@encryptDir} --storekey")
    rp = build_repo_obj
    rp.open(@@sourceDir)
    rp.perform_update(true)
    rp.close  
  end

  def test_110_open_repo_using_Repo_class
    # (assumes repo exists at this point; preceding test creates it)
    rp = build_repo_obj
    rp.open(@@sourceDir)
  end

  def test_120_update_repo_using_Repo_class
    update_repo
    assert(File.directory?(@@encryptDir))
  end
  
  def test_121_update_repo_with_lencrypt_program  
    remove_file_or_dir(@@encryptDir)
    ex("")
    assert(File.directory?(@@encryptDir))
  end

  def test_125_recover_repository
    remove_file_or_dir(@@recoverDir)
    do_recover
    assert(File.directory?(@@recoverDir))  
  end
  
  def test_130_update_when_repository_not_found  
    assert_raise(RepoNotFoundException) do 
      clean
      rp = build_repo_obj
      rp.open(@@sourceDir)
    end
  end
  
  
  def test_200_recover_with_incorrect_password   
    create_repo
    ex("") # updates repo
    assert_raise(DecryptionError) do
      remove_file_or_dir(@@recoverDir)
      rp = build_repo_obj()
      rp.perform_recovery("thisisthewrongkey_______", @@encryptDir,@@recoverDir)
      rp.close
    end
  end
  
end


lib/lenc/lencrypt.rb
require_relative 'repo'

# The application script (i.e., the 'main program')
#
class LEncApp
  include LEnc
  
  def run(argv = ARGV) 
     
    req 'trollop'
    p = Trollop::Parser.new do
        opt :init, "create new encryption repository: KEY ENCDIR ", :type => :strings
        opt :orignames, "(with --init) leave filenames unencrypted"
        opt :storekey, "(with --init) store the key within the repository configuration file so it" \
              " need not be entered with every update"
        opt :update, "update encrypted repository (default operation): KEY", :default => ""
        #opt :updatepwd, "specify key for update: KEY", :type => :string
        opt :recover, "recover files from an encrypted repository: KEY ENCDIR RECDIR", :type => :strings
        opt :where, "specify source directory (default = current directory)", :type => :string
        opt :verbose,"verbose operation"
        opt :quiet, "quiet operation"
        opt :dryrun, "show which files will be modified, but make no changes"
    end
    
    options = Trollop::with_standard_exception_handling p do
      p.parse argv
    end
    
    # Not sure how to determine if there were leftover arguments;
    # trollop seems to include path information ('.')
    #p.die("Unrecognized argument: #{p.leftovers[0]}",nil) if p.leftovers.size
    
    v = 0
    v = -1 if options[:quiet]
    v = 1  if options[:verbose]
    
    update_pwd = options[:update]
    update_pwd = nil if update_pwd.size == 0
    
    nOpt = 0
    nOpt += 1 if options[:init]
    nOpt += 1 if update_pwd  
    nOpt += 1 if options[:recover]
            
    #pr("trollop opts = %s\n",d2(options))
      
    p.die("Only one operation can be performed at a time.",nil) if nOpt > 1
  
    r = Repo.new(:dryrun => options[:dryrun],
      :verbosity => v)
    
    begin

      if (a = options[:init])
        p.die("Expecting: KEY ENCDIR",nil) if a.size != 2
        pwd,encDir = a
        r.create(options[:where], pwd, encDir, options[:orignames], options[:storekey])
      elsif (a = options[:recover])
        p.Trollop::die("Expecting: KEY ENCDIR RECDIR",nil) if a.size != 3
        r.perform_recovery(a[0],a[1],a[2])
      else
        r.open(options[:where],update_pwd)
        r.perform_update(options[:verifyenc])
      end

      r.close()
    rescue Exception =>e
      puts("\nProblem encountered: #{e.message}")
    end

  end
end

if __FILE__ == $0
  args = ARGV
  
  
  LEncApp.new().run(args)
end
  


lib/lenc/config_file.rb
require 'json'
require_relative 'tools'

module LEnc

  # Manages a configuration file, which
  # is a set of key/value pairs that can be saved to the file system.
  #
  class ConfigFile 
  
    attr_reader :path
    
    # @param filename where configuration file is to be found (or written to,
    #    if it doesn't yet exist)
    # @param parentDir directory containing file; if nil, uses user's home directory
    #
    def initialize(filename, parentDir=nil) 
      
      parentDir ||= Dir.home
        
      @path = File.join(parentDir, filename)
      
      @content = {}
      @origContentStr = "!!!"
      if exists()
        contents = read_text_file(@path).strip
        
        @content = JSON.parse(contents)
        @origContentStr = JSON.dump(@content)
      end
    end
  
    # Determine if configuration file exists on disk
    def exists()
      File.exists?(@path)
    end
  
    # Get the directory containing the configuration file
    def get_directory()
      File.dirname(@path)
    end
    
    # Store a key => value pair (any existing value for this key is overwritten)
    def set(key,val)
      @content[key] = val
    end
    
    # Remove key (and value), if it exists
    def remove(key)
      @content.remove(key)
    end

    # Write configuration file, if it has changed
    def write
      newStr = JSON.dump(@content)
    
      if newStr != @origContentStr 
        @origContentStr = newStr
        write_text_file(@path,@origContentStr + "\n")
      end
    end
     
    # Get value for a key
    # @param defVal value to return if key doesn't exist
    def val(key, defVal = nil)
      @content[key] || defVal
    end
      
    def to_s
      s = 'ConfigFile ' 
      s << path()
      s << dh(@content)
#      " [\n"
#      @content.each_pair do |k,v|
#        s << d(k) << ' ==> ' << d(v) << "\n"
#      end
#      s << "]\n"
      s 
    end
    
    def inspect
      to_s
    end  
    
  end # Class
end # Module


if main? __FILE__

  include LEnc
  
  f = ConfigFile.new("__testrbconfig__.txt")
  
  pr("constructed:\n%s\n",d(f))
  
  srand
  "alpha bravo echo geronimo".split.each do |s|
    n = s.size
    if rand() > 0.8
      n = rand(12)
    end
    f.set(s,n)
  end
  
  f.write
  
end


lib/lenc/aes.rb
require_relative 'tools'

require 'openssl'


module RepoInternal
  
# --------------------------------------------------------------

  AES_BLOCK_SIZE = 16
  PAD_BYTE = 254
  PAD_CHAR = PAD_BYTE.chr
  
  CHUNK_HEADER_SIZE = 8
  CHUNK_VERIFY_SIZE = CHUNK_HEADER_SIZE - 1
  CHUNK_VERIFY_STR = 0.chr * CHUNK_VERIFY_SIZE
  
  # Size of input chunks during decryption (they include space for a header);
  # must be a multiple of AES_BLOCK_SIZE
  CHUNK_SIZE_DECR = 1 << 16
  
  # Size of input chunks during encryption
  CHUNK_SIZE_ENCR = CHUNK_SIZE_DECR - CHUNK_HEADER_SIZE
  
  # The size of nonce the underlying API expects
  NONCE_SIZE_LARGE = 16
      
  # The size of nonce we'll be using (we'll pad it out with zeros 
  # when a full size one is required)
  NONCE_SIZE_SMALL = 8
  
 
    
=begin
    Wrapper for OpenSSL AES cipher.
    
  
    Usage for encryption:
    ----------------------------------------------------
    original = "..."          # bytes to encrypt (string)
    
    key = "xxxx..."           # encryption key (string of size 8..56)
    
    en = MyAES.new(true, key) # construct an encryptor    
    
    en.finish(original)       # add data to encrypt          
    
    encrypted = en.flush()    # get encrypted bytes (string)
    ----------------------------------------------------
    
  
      
    Usage for decryption:
    ----------------------------------------------------
    encrypted = "..."         # bytes to decrypt 
    
    key = "xxxx...."    
    
    de = MyAES.new(false, key)  # construct a decryptor
    
    de.finish(original)       # add data to decrypt
    
    decrypted = de.flush()    # get decrypted bytes
    
    ----------------------------------------------------
      
  
    Use of nonces:
    --------------
    The above encryption example generates a new (hopefully unique) 'nonce' 
    which is an added security feature.  It uses the system clock to do this.
    This means the same file will produce different encrypted byte streams on 
    repeated encryption attempts, which may be undesirable.  A fixed nonce
    (for a particular input file) can be specified as an additional input:
  
    nonce = "nnnn.."         # only the first 8 bytes are used
    en = MyAES.new(true, key, nonce)
    
    The nonce, whether explicitly given or randomly generated, is added to the
    encrypted stream; hence it need not be specified when decrypting.
    
    Stream mode:
    ------------
    When processing large files, you may want to do them a chunk at a time.
    Here's an example of encrypting using stream mode (decrypting is similar):
    
    
    en = MyAES.new(true, key)    
    
    s = {size of input file}
    n = 0
    while n < s  
      c = [5000, s - n].min
      en.add( {bytes n..n+c-1 from the input file} )
      r = en.flush()
      {append bytes r to output file}
    end
  
    en.finish()
    r = en.flush()
    {append bytes r to output file}
    
    ----------------------------------------------------
     
    Format of encrypted data:
    
    [8] nonce (only the first 8 bytes of the nonce are actually used)
    
    Followed by one or more encrypted chunks of length [k], where k is 65536, unless it's the last
    chunk in the file, in which case it must be a multiple of 16.
  
    The first bytes of each decrypted chunk is a header:
      [7]  zeros
      [1]  number of padding bytes present at end of block
       
       
    For example, suppose a file of 71980 'source' bytes has been encrypted.  The encrypted file will contain:
      [8] nonce
      [65536] first chunk, consisting of
         [7] zeros
         [1] zero, since this chunk needed no padding
         [65528] 65528 encrypted source bytes
      [6464]  second chunk, consisting of
         [7] zeros
         [1] 4, indicating 4 padding bytes
         [6456] 6452 encrypted source bytes plus 4 padding bytes 
         
      Observe that 65528 + 6452 = 71980.
    
    The purpose of the [7] zeros in the (decrypted) chunk header are to indicate
    whether decryption was successful (e.g., if the password was correct).  The assumption
    is that an incorrect password will generate 7 zeros in these locations with extremely low probability.
       
    The byte used as a padding byte is 254.
    
    If a file has length zero, then when encrypted, it will have the following structure:
      [8] nonce
      [16] chunk:
        [7] zeros
        [1] 8, indicating 8 padding bytes
        [8] 0 encrypted source bytes plus 8 padding bytes
      
=end
  class MyAES 
    
    private
    
    # decryptState values
    DS_WAITNONCE = 0 # waiting for nonce to appear in input
    DS_WAITCHUNK = 1 # waiting for encrypted chunk to appear 
    
    # A class variable that increments with each encryptor object constructed,
    # to help generate unique nonces (in conjunction with system clock)
    @@nonceHelper = 0
      
    # Construct a MyAES object to encrypt/decrypt a sequence of bytes.
    # @param encrypting true for encryption, false for decryption
    # @param key   a bytearray of 4..56 bytes
    # @param nonce  a string of up to 16 characters; if nil, one is 
    #   generated from the system clock
    def initialize(encrypting, key, nonce=nil)
      
      @encrypting = encrypting
      @inputBuffer = ''
      @outputBuffer = ''
      @finished = false
      
      if nonce && !encrypting
        raise ArgumentError, \
         "nonce should not be supplied during decryption"
      end 
      
      if @encrypting
        @nonceWritten = false
      else 
        @decryptState = DS_WAITNONCE
      end
      
      key = bytes_to_str(key)
        
      if key.size < 4 || key.size > 56
          raise ArgumentError, 'Key length not 4..56 bytes' 
      end
        
      # expand the key to be at least 32 bytes
      k = (32.0 / key.size).ceil.to_i
      key = key * k
      
      @key = key[0...32]
      @chunkCount = 0
      
      # If we are encrypting, set nonce; otherwise, we must wait for some data to be available
      if @encrypting 
        setNonce(nonce) 
      else
        @chunkExpected = true
      end
    end
    
    # Set nonce
    # @param nonce a string; if nil, uses system clock and an internal
    #   counter to generate (hopefully) a unique value
    def setNonce(nonce=nil)
      if !nonce
          # use date-time if no counter provided
          ni = int_to_bytes(Time.now.usec)
          ni.concat(int_to_bytes(@@nonceHelper))
          @@nonceHelper += 1
          nonce = bytes_to_str(ni)
      end
      
      raise ArgumentError if !(nonce.is_a? String )
      simple_str(nonce)
      
      
      nonce = str_sized(nonce,NONCE_SIZE_SMALL)
      
      @nonce = nonce
    end  
  
    def incrNonce()
      c = @nonce
      dig = NONCE_SIZE_SMALL - 1
      while true
        raise ArgumentError, "Nonce overflow" if dig < 0
        
        q = c[dig].ord
        if q != 0xff 
          c[dig] = (q+1).chr
          break
        end
        
        c[dig] = 0.chr
        dig -= 1
      end
    end
    

    # Convert nonce from our version to one the OpenSSL expects.
    # This may involve padding or truncating it as necessary to a
    # fixed length.
    def cvtNonce()
      str_sized(@nonce,NONCE_SIZE_LARGE)
    end
    
    def buildAES()
      aes = OpenSSL::Cipher.new("AES-256-CBC")
      aes.padding = 0  # Not sure this is required, as we are doing padding ourselves;
                       # we should only be asking it to process blocks that need no padding
      aes
    end
    
    def processChunk()
  
      if @encrypting 
          
        csize = [CHUNK_SIZE_ENCR, @inputBuffer.size].min
        
        padBytes = (-(csize + CHUNK_HEADER_SIZE)) & (AES_BLOCK_SIZE - 1)
        
        csize += padBytes
        
        if padBytes 
          @inputBuffer << PAD_CHAR * padBytes
        end
        
        aes = buildAES()
        aes.encrypt
        aes.key = @key
        nonceStr = cvtNonce
        aes.iv = nonceStr
        
        if not @nonceWritten 
          @nonceWritten = true
          @outputBuffer  << nonceStr[0...NONCE_SIZE_SMALL]
        end
                  
        cdata = "\0" * CHUNK_VERIFY_SIZE
        cdata << padBytes.chr
        cdata << @inputBuffer.slice!(0,csize)  
          
        @outputBuffer << aes.update(cdata)
        @outputBuffer << aes.final
        
      else  # Decrypting
        
        csize = [CHUNK_SIZE_DECR,@inputBuffer.size].min
        
        # verify that the chunk size is a nonzero multiple of AES_BLOCK_SIZE bytes
        if not csize or 0 != (csize & (AES_BLOCK_SIZE - 1)) 
          raise LEnc::DecryptionError, "chunk size not a multiple of block size" 
        end
        
        aes = buildAES()
        aes.decrypt
        aes.key = @key
        aes.iv = cvtNonce()
        
        cdata = @inputBuffer[0...csize]
        newData = aes.update(cdata) 
        newData << aes.final

        if !newData.start_with? CHUNK_VERIFY_STR
          raise LEnc::DecryptionError, "header doesn't verify"
        end
          
        nPadBytes = newData[CHUNK_VERIFY_SIZE].ord  
        actualEnd = csize - nPadBytes
        if nPadBytes > 16 or actualEnd < CHUNK_HEADER_SIZE 
          raise LEnc::DecryptionError, "nPadBytes/actualEnd mismatch"
        end  
        
        

        # Verify that the padding bytes have correct values
        (actualEnd...csize).each do |i|
          if newData[i] != PAD_CHAR
            raise LEnc::DecryptionError,"padding char bad value"
          end
        end
          
        newData = newData[CHUNK_HEADER_SIZE ... actualEnd]
          
        @decryptState = DS_WAITCHUNK
          
        @inputBuffer.slice!(0,csize)
        @outputBuffer  << newData
        
      end
      
      incrNonce()
      @chunkCount += 1
    end
    
    public
     
    # Process additional input bytes, encrypting (or decrypting) its contents
    # @param data string containing input bytes
    def add(data)
       
      raise IllegalStateException if @finished 
      
      simple_str(data)
      
      @inputBuffer << data #.concat(data)
    
      while true 
    
        if not @encrypting 
          
          # Extract nonce if we're waiting for it and it is now available
          if @decryptState == DS_WAITNONCE 
            break if @inputBuffer.size <  NONCE_SIZE_SMALL 
            setNonce(@inputBuffer.slice!(0...NONCE_SIZE_SMALL))  
            @decryptState = DS_WAITCHUNK
            next
          end
          
          # If we don't have a full chunk, exit
          # (the last chunk may be smaller; we'll test for this when finishing up)
          break if @inputBuffer.size < CHUNK_SIZE_DECR
          
        else 
          break if @inputBuffer.size < CHUNK_SIZE_ENCR
        end  
        
        # Process chunk and repeat
        processChunk()
      end
    end

    # Stop the encryption/decryption process.  
    # Processes any bytes that may have been buffered (since encryption occurs in 
    # 16 byte blocks at a time).
    # 
    # @param data   optional final input string to process before finishing
    # 
    def finish(data = nil) 
      
      add(data) if data 
        
      raise IllegalStateException if @finished  
      
      @finished = true
      
      inpLen = @inputBuffer.size
      
      if @encrypting 
        # If input buffer is not empty, or we haven't written a first chunk (which contains the nonce),
        # encrypt a chunk
        if inpLen or (not @nonceWritten) 
          processChunk()
        end
      else
        
        # We must be at WAITCHUNK with an input buffer that is a multiple of _AES_BLOCK_SIZE bytes in length
        if @decryptState != DS_WAITCHUNK or 0 != (inpLen & (AES_BLOCK_SIZE-1)) 
          raise LEnc::DecryptionError, "decrypt state problem"
        end
        
        # We expect a chunk if there's more input, or if we've never processed a chunk.
        if inpLen != 0 or @chunkCount == 0
          processChunk()
        end
      end
    end
        
      
    # Return any output bytes that have been generated since the last call to flush()
    # @return string containing bytes
    def flush()
      ret = @outputBuffer 
      @outputBuffer = ''
      return ret
    end

    # Strip the header from an encrypted string
    def strip_encryption_header( encr_str)
      return encr_str[CHUNK_HEADER_SIZE..-1]
    end
        
    # Determines if a string is the start of an encrypted sequence
    # 
    # @param key  password to use (string) 
    # @param test_str the string to test
    #  
    # Returns true iff the start of the string seems to decrypt correctly
    # for the given password
    def self.is_string_encrypted(key, test_str) 
      db = warndb 0
      
      !db || hex_dump(test_str, "areBytesEncrypted?")
      
      simple_str(test_str)
      
      lnth = test_str.size
      lnth -= NONCE_SIZE_SMALL
      if lnth < AES_BLOCK_SIZE 
        !db || pr("  insufficient # bytes\n")
        return false
      end
      
      begin
          de = MyAES.new(false, key)    
          de.finish(test_str[0...AES_BLOCK_SIZE + NONCE_SIZE_SMALL])
          decr = de.flush()
          !db || hex_dump(decr,"decrypted successfully")
      rescue LEnc::DecryptionError
        !db || pr(" (caught DecryptionError)\n")
        return false
      end
        
      true
    end
      
    # Determines if a file is an encrypted file
    # @param key    password to use (string, or array of bytes) 
    # @param path    path to file
    # @return true iff the start of the file seems to decrypt correctly
    # for the given password, and the file is of the expected length.
    def self.is_file_encrypted(key, path) 
      
    #    key = str_to_bytes(key)
      
      if not File.file?(path) 
        return false
      end
      
      lnth = File.size(path)
      minSize = NONCE_SIZE_SMALL + AES_BLOCK_SIZE
      if lnth < minSize or ((lnth - minSize) % _AES_BLOCK_SIZE) != 0
        return false
      end
      
      f = File.open(path,"rb")
      return is_string_encrypted(key, f.read(minSize))
    end
  
  
  end # end of class MyAES

end # module RepoInternal


if main? __FILE__
  
  s = ''
  16.times {|x| s << (65+x).chr}
    

  nonce  = "abc" * 20
  nonce  = nonce[0...16]
  key = "onefishtwofishredfishbluefish" * 3
  key = key[0...32]
  
  hex_dump(key,"key")
  hex_dump(nonce,"nonce")
  
  aes = OpenSSL::Cipher.new("AES-256-CBC")
      
  aes.padding = 0
  aes.encrypt
  aes.key = key
        
  aes.iv = nonce
        
  enc = aes.update(s)
  enc << aes.final
        
  hex_dump(s,"calling aes.encrypt with")
  hex_dump(enc,"aes.encrypt returned")

  s = enc
  
  require 'base64'
  s = Base64.urlsafe_encode64(s)
  hex_dump(s,"base64")

end